### ISSUE TITLE
Depth from Videon in the Wild fails with NaN due to CUDA bug

### ISSUE THREAD

**Issue by fxtentacle** [google-research/google-research#162]
It appears that 
https://github.com/google-research/google-research/blob/b13a2d98f590495331e27b86268b000413ecb50d/depth_from_video_in_the_wild/transform_utils.py#L111
can produce arbitrarily wrong gradients on trans_vec1 which then back-propagate to corrupt the depth estimation network with NaN weights. 

The issue is caused by a `CUDA 10.0` bug https://github.com/tensorflow/tensorflow/issues/31166 which is fixed in `TF 2 nightly` by upgrading to `CUDA 10.1`, however this network requires `TF 1.x`. One workaround would be to recompile `TF 1.15` with `CUDA 10.1` or to use CPU for all calculations.

But in any case, the network as-is won't train on GPU for more than a few steps before going NaN with any released TF version. So I'm writing this issue mainly to document that for others.

**Comment by pigletsc** [comment]
@fxtentacle 
Hi, Are you talking about a error message like "LossTensor is inf or nan : Tensor had NaN values" when training from scratch?

Correct me if I am wrong
Thanks

**Comment by fxtentacle** [comment]
@pigletsc  Yes, except that the error occurs not only during training from scratch, but also during inferrence and/or when one continues training the pre-trained model for a few steps.

**Comment by asherRSIP** [comment]
any progress?
I also get: "LossTensor is inf or nan : Tensor had NaN values" message when continuing to train the pre-trained model.

**Comment by pigletsc** [comment]
@fxtentacle Thanks for reply!
@asherRSIP After tracing the code and print all loss terms, I think @fxtentacle is right. 
If I train the model by CPU, there is no Nan errors.
I will try to train on GPU by CUDA 10.1, and if there are some results, I will post them here.
 btw, @asherRSIP what is your pre-trained model? Do you mean resnet-18 which train on ImageNet?
because I can not load the ImageNet pre-trained model, so all I can do is train from scratch.

Thanks

**Comment by vibudh2209** [comment]
@pigletsc I am facing the same issue, did it work with CUDA 10.1?

**Comment by fxtentacle** [comment]
@vibudh2209  I compiled TensorFlow 2.0 with VS2017 and CUDA10.1 on Windows. It took 2 hours and 30+ GB of hard disk space, and now the model works without NaNs.

That said, I am also seeing gradient propagation problems, so training this model from scratch probably involves some luck in dropping into the right local minima upon initialization, in addition to recompiling TF.

**Comment by vibudh2209** [comment]
@fxtentacle thanks for the info, is it possible for you to share your tensorflow wheel with me? 
Do you have to make a lot of changes to the code for switching to tensorflow 2.0?

**Comment by fxtentacle** [comment]
@vibudh2209 I had to make a lot of changes because `tf.contrib.image.transform` was dropped from TF1 to TF2, so you basically need to re-implement parts from scratch. And if you get there, the mathematical issues kick in and make learning expensive or impossible. So in the end, I abandoned my quest to make the paper work, because it did not seem worthwhile, even after the crash, NaN and TF2 incompatibility issues were solved.

**Comment by vibudh2209** [comment]
@fxtentacle I see, that's sad. Just one last thing, do you think it's worthwhile to compile tensorflow 1.15 with CUDA 10.1 instead (since it should not require to change the code substantially)?

**Comment by vibudh2209** [comment]
So building tensorfow 1.15.2 with CUDA 10.1 solved the problem for me (without changing any code).

**Comment by Truantboy** [comment]
As @fxtentacle said, it's a bug in CUDA10.0 when using tf.matmul. I did some test and found that this error only happens when the matrix is too big, so we can separate the large matrix to small matrixs and then merge the result.
Here is a solution without using CUDA10.1
```
def tf_matmul_separate(A,B):
    batch,a_h,a_w,_,_ = A.get_shape()
    if batch*a_h*a_w < 80000:
        return tf.matmul(A,B)
    w_half = tf.cast(tf.div(a_w,2),tf.int32)
    h_half = tf.cast(tf.div(a_h,2),tf.int32)
    A_1 = A[:,:h_half,:w_half,:,:]
    A_2 = A[:,:h_half,w_half:,:,:]
    A_3 = A[:,h_half:,:w_half,:,:]
    A_4 = A[:,h_half:,w_half:,:,:]
    
    B_1 = B[:,:h_half,:w_half,:,:]
    B_2 = B[:,:h_half,w_half:,:,:]
    B_3 = B[:,h_half:,:w_half,:,:]
    B_4 = B[:,h_half:,w_half:,:,:]
    
    res_1 = tf_matmul_separate(A_1,B_1)
    res_2 = tf_matmul_separate(A_2,B_2)
    res_3 = tf_matmul_separate(A_3,B_3)
    res_4 = tf_matmul_separate(A_4,B_4)
    tmp_up = tf.concat([res_1,res_2],2)
    tmp_down = tf.concat([res_3,res_4],2)
    final_res = tf.concat([tmp_up,tmp_down],1)
    return final_res
```
using this function replace the tf.matmul function in:
https://github.com/google-research/google-research/blob/master/depth_from_video_in_the_wild/transform_utils.py#L110
https://github.com/google-research/google-research/blob/master/depth_from_video_in_the_wild/transform_utils.py#L111

**Comment by jessedaniels** [comment]
Another workaround for this problem without upgrading CUDA is to increase the precision of the tensors from float32 to float64, then cast back to float32 after matmul:

`rot_mat1 = tf.cast(rot_mat1, tf.float64)
  trans_vec1 = tf.cast(trans_vec1, tf.float64)
  rot_mat2 = tf.cast(rot_mat2, tf.float64)

  r2r1 = tf.matmul(rot_mat2, rot_mat1)
  r2t1 = tf.matmul(rot_mat2, tf.expand_dims(trans_vec1, -1))
  r2t1 = tf.squeeze(r2t1, axis=-1)

  r2r1 = tf.cast(r2r1, tf.float32)
  r2t1 = tf.cast(r2t1, tf.float32)

  return r2r1, r2t1 + trans_vec2`

**Comment by BARAKRAPAT** [comment]
Hi, when I try to train the network (in depth_and_motion_learning project), I get NaN in loss function even when I train on CPU. 
The error is:
"tensorflow.python.training.basic_session_run_hooks.NanLossDuringTrainingError: NaN loss during training."
does anybody now why it occurs?
